### PR TITLE
Event details

### PR DESCRIPTION
**Pull request by farisashai**
# Description

- Uses portal events API to generate dynamic path urls statically in the form /events/[name]-[uuid] for each existing event (works for past events too)
- Event card on /events links to event page
- Implemented event page design with mobile responsiveness
- Feature: Save details to calendar by 
  - Download as ICS
  - Save to Google Cal link
- Add event link previews to optimize SEO

*TODO*
- Not sure how we should implement the design for the calendar feature. ronak suggested a dropdown menu on click but i'm hesitant to install ant design or an entire library just for that one item.
- Will be styling mobile dropdown on /events separately and then i think we're done
- should we implement a feature to view past events on /events? it's already supported just by typing the uuid of a past event in the url and will generate the page correctly. 

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Logistical Change (something about code process / a readme change / documentation update)
- [ ] CI Change (relating to deployment / continuous integration)
- [ ] Something Else 

# How Has This Been Tested?
- [x] Tested on Localhost
- [x] Tested on Deployed Preview

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code as needed, particularly in hard-to-understand areas
- [x] My changes generate no new warnings (if they do, please explain why)

### CODE REVIEW

**Comment by trulyronak** [comment]
Functionally looks good, will do a code review in a bit

Addressing your TODOs:

> Not sure how we should implement the design for the calendar feature. ronak suggested a dropdown menu on click but i'm hesitant to install ant design or an entire library just for that one item.
- we shouldn't have to get a library to implement dropdowns, but I think your temporary fix for now is fine - we'll work on potentially getting a designer involved next quarter to patch up some of the remaining stpes

> should we implement a feature to view past events on /events? it's already supported just by typing the uuid of a past event in the url and will generate the page correctly.
- adding the support early is perfect, we still need to design a system for what it will look like; for now, keeping that functionality as is is perfect

some questions - 

- if someone creates a portal event today, how soon will it take for it to propagate over the main site? it seems like you mentioned it all happens dynamically, so it should be instant right
- what willt he media link preview look like (the embed) → can we make it look nicer (event name / date / graphic)

File: src/api/EventsAPI.ts [hunk 1]
```diff
@@ -19,30 +16,101 @@ export type EventObject = {
...
   requiresStaff: boolean;
   staffPointBonus: number;
+  facebookUrl?: string;
 };
 
-const handleErrors = (response: Response): Promise<EventsResponse> => {
+export type EventsArray = EventObject[];
+
+export type EventsResponse = {
+  error: unknown;
```
Comment (trulyronak) [inline src/api/EventsAPI.ts:25]: why is this type `unknown`? what shape does it actually end up taking (@StormFireFox1 is there a standardized error type we can expect to get from portal)

File: src/api/EventsAPI.ts [hunk 2]
```diff
@@ -19,30 +16,101 @@ export type EventObject = {
...
+    const response: any = await fetch(api_url);
+    const result: EventResponse = await handleErrors(response);
+    return result.event;
+  } catch (error) {
+    return undefined;
+  }
+};
+
+const createDownloadFile = (textContent: string, title: string): void => {
+  const element = document.createElement("a");
```
Comment (trulyronak) [inline src/api/EventsAPI.ts:68]: credit the stackoverflow where we lifted this from so we can maintain it better in the future

File: src/api/EventsAPI.ts [hunk 3]
```diff
@@ -19,30 +16,101 @@ export type EventObject = {
...
+
+const downloadICS = (event: EventObject): void => {
+  const startDate = new Date(event.start);
+  const endDate = new Date(event.end);
+  const duration = endDate.getTime() - startDate.getTime();
+
+  const attributes: ics.EventAttributes = {
+    start: [
+      startDate.getFullYear(),
+      startDate.getMonth() + 1,
```
Comment (trulyronak) [inline src/api/EventsAPI.ts:84]: why do we need to add one?

File: src/api/EventsAPI.ts [hunk 4]
```diff
@@ -19,30 +16,101 @@ export type EventObject = {
...
+    organizer: { name: "ACM at UCSD", email: "acm@ucsd.edu" },
+  };
+
+  const response = ics.createEvent(attributes);
+
+  createDownloadFile(response.value, formatURLEventTitle(event.title));
+};
+
+const saveToGoogleCal = ({ title, description, location, start, end }: EventObject): void => {
+  const url = new URL("https://www.google.com/calendar/render?action=TEMPLATE");
```
Comment (trulyronak) [inline src/api/EventsAPI.ts:103]: link to the template we're using for this; there's always a chance google deprecates this so it'll be good to know where we got this pattern from (if you can find an api reference that's even better)

File: src/components/EventCard/index.tsx [hunk 5]
```diff
@@ -9,36 +11,39 @@ const EventCard: React.FC<{
...
+        <h2 className={s.cardDay}>{day}</h2>
+      </div>
+      <div className={s.cardBody}>
+        <h3 className={`${s.eventTitle} ${s[org.toLowerCase()]}`}>{title}</h3>
+        <h3 className={s.eventLocation}>{location}</h3>
+        <h3 className={s.eventTime}>{time}</h3>
+      </div>
+      <div className={s.cardFooter}>
+        <a role="button">
+          <img src="/assets/calendar.svg" alt="" width={20} height={20} className={s.footerIcon} />
```
Comment (trulyronak) [inline src/components/EventCard/index.tsx:32]: out of curiosity, why don't we use the next image component?

File: src/components/NavigationBar/index.tsx [hunk 6]
```diff
@@ -21,7 +21,7 @@ const NavigationBar: React.FC = () => {
 
   // Switch to mobile less than 920px
```
Comment (trulyronak) [inline src/components/NavigationBar/index.tsx:22]: update this comment if we're changing it

File: src/sections/event/Event.Content.tsx [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+  const date = new Date(event.start).getDate();
+  const day = days[(new Date(event.start).getDay() + 6) % 7];
+  const time = getDateTime(event).time;
+  const { committee, title, location, description, start, end, facebookUrl } = event;
+
+  return (
+    <div className={s.container}>
+      <div className={s.eventHeader}>
+        <Link href={"/events"}>
+          <a className={s.backButton}>&lt; Back to Events</a>
```
Comment (trulyronak) [inline src/sections/event/Event.Content.tsx:17]: we should be using the next link component for all a tags

**Review by trulyronak — CHANGES_REQUESTED** [review]

File: src/api/EventsAPI.ts (hunk 3 above)
Comment (farisashai) [reply src/api/EventsAPI.ts:84]: js dates are zero indexed but the library is 1 indexed

File: src/components/EventCard/index.tsx (hunk 5 above)
Comment (farisashai) [reply src/components/EventCard/index.tsx:32]: next images are rly hard to style but i can definitely change it here since it's a fixed size

Comment (trulyronak) [reply src/components/EventCard/index.tsx:32]: specify this in a comment so it makes sense for future maintenance; that's an interesting but i guess logical choice by the library

File: src/api/EventsAPI.ts (hunk 3 above)
Comment (trulyronak) [reply src/api/EventsAPI.ts:84]: specify this in a comment so it makes sense for future maintenance; that's an interesting but i guess logical choice by the library

**Comment by farisashai** [comment]
so the bug i was running into is that it turns out we can add static paths once at build time by fetching all future events but nextjs doesn't support revalidation to add new event paths every hour. we can only rebuild existing pages with new props using incremental static regeneration. this means that all dynamic event routes have to be built server side (not statically) when the routes are fetched for the first time (caching them for later visits) so it might not be possible to do seo previews. 

i'm still testing this and trying to find workarounds.

File: src/api/EventsAPI.ts (hunk 1 above)
Comment (farisashai) [reply src/api/EventsAPI.ts:25]: from what I can tell, it's impossible to throw an error in the first place but idk?
 
Screen Shot 2022-03-24 at 12 44 28 AM
e

**Review by trulyronak — APPROVED** [review]
good stuff

File: src/api/EventsAPI.ts (hunk 1 above)
Comment (shravanhariharan2) [reply src/api/EventsAPI.ts:25]: We will be publishing our error type during the workspace refactor on portal backend, so you can use that here once that's done. It is possible for errors to be thrown here, e.g. if an invalid query parameter is supplied, or if there's some internal database error, so we can't rule that out entirely (though realistically this route shouldn't throw many errors at all)

**Pull request merged**